### PR TITLE
feat: implement lgamma, algama, dlgama

### PR DESCRIPTION
**Pull request by parth121101**
Towards : #492

### CODE REVIEW

**Comment by parth121101** [comment]
@Shaikh-Ubaid can you please try this PR on your system. Its working fine on my device but not able to figure out why the CI is failing.

**Comment by Shaikh-Ubaid** [comment]
>@Shaikh-Ubaid can you please try this PR on your system. Its working fine on my device but not able to figure out why the CI is failing.

It works on my system (Mac M1) locally. It passes for Mac at the CI. It fails on Linux systems at the CI. What system do you use @parth121101 ? I think we need to test/check this on Linux.

**Comment by parth121101** [comment]
> >@Shaikh-Ubaid can you please try this PR on your system. Its working fine on my device but not able to figure out why the CI is failing.
> 
> It works on my system (Mac M1) locally. It passes for Mac at the CI. It fails on Linux systems at the CI. What system do you use @parth121101 ? I think we need to test/check this on Linux.

Even I am using Mac

**Comment by Shaikh-Ubaid** [comment]
I don't have access to a Linux computer at the moment. We need someone who has a Linux system to help us with the error. @HarshitaKalani do you use a linux system? If yes, can you please help us in reproducing the error or creating an MRE for it?

**Comment by HarshitaKalani** [comment]
```fortran
program intrinsics_231
    use, intrinsic :: iso_fortran_env, only: dp => real64, sp => real32
    real(sp) :: y, z
    real(dp) :: a, b, c

    y = 23.82920_sp
    z = 62.739713_sp

    a = 9271.72826260_dp
    b = 23.82926260_dp
    c = 62.73971326260_dp

    print *, lgamma(y)
    if (abs(lgamma(y) - 5.10680733e+01) > 1e-5) error stop

    print *, lgamma(23.82920)
    if (abs(lgamma(23.82920) - 5.10680733e+01) > 1e-5) error stop

    print *, lgamma(z)
    ! if (abs(lgamma(z) - 1.95790390e+02) > 1e-5) error stop

    print *, lgamma(62.739713)
    if (abs(lgamma(62.739713) - 1.95790390e+02) > 1e-5) error stop

    print *, lgamma(a)
    if (abs(lgamma(a) - 7.54193119924598432e+04_dp) > 1e-12) error stop

    print *, lgamma(9271.72826260_dp)
    if (abs(lgamma(9271.72826260_dp) - 7.54193119924598432e+04_dp) > 1e-12) error stop

    print *, lgamma(b)
    if (abs(lgamma(b) - 5.10682662991165941e+01_dp) > 1e-12) error stop

    print *, lgamma(23.82926260_dp)
    if (abs(lgamma(23.82926260_dp) - 5.10682662991165941e+01_dp) > 1e-12) error stop

    print *, lgamma(c)
    if (abs(lgamma(c) - 1.95790392620801725e+02_dp) > 1e-12) error stop

    print *, lgamma(62.73971326260_dp)
    if (abs(lgamma(62.73971326260_dp) - 1.95790392620801725e+02_dp) > 1e-12) error stop

    print *, algama(y)
    if (abs(algama(y) - 5.10680733e+01) > 1e-5) error stop

    print *, algama(23.82920)
    if (abs(algama(23.82920) - 5.10680733e+01) > 1e-5) error stop

    print *, algama(z)
    ! if (abs(algama(z) - 1.95790390e+02) > 1e-5) error stop

    print *, algama(62.739713)
    if (abs(algama(62.739713) - 1.95790390e+02) > 1e-5) error stop

    print *, algama(a)
    if (abs(algama(a) - 7.54193119924598432e+04_dp) > 1e-12) error stop

    print *, algama(9271.72826260_dp)
    if (abs(algama(9271.72826260_dp) - 7.54193119924598432e+04_dp) > 1e-12) error stop

    print *, algama(b)
    if (abs(algama(b) - 5.10682662991165941e+01_dp) > 1e-12) error stop

    print *, algama(23.82926260_dp)
    if (abs(algama(23.82926260_dp) - 5.10682662991165941e+01_dp) > 1e-12) error stop

    print *, algama(c)
    if (abs(algama(c) - 1.95790392620801725e+02_dp) > 1e-12) error stop

    print *, algama(62.73971326260_dp)
    if (abs(algama(62.73971326260_dp) - 1.95790392620801725e+02_dp) > 1e-12) error stop

    print *, dlgama(a)
    if (abs(dlgama(a) - 7.54193119924598432e+04_dp) > 1e-12) error stop

    print *, dlgama(9271.72826260_dp)
    if (abs(dlgama(9271.72826260_dp) - 7.54193119924598432e+04_dp) > 1e-12) error stop

    print *, dlgama(b)
    if (abs(dlgama(b) - 5.10682662991165941e+01_dp) > 1e-12) error stop

    print *, dlgama(23.82926260_dp)
    if (abs(dlgama(23.82926260_dp) - 5.10682662991165941e+01_dp) > 1e-12) error stop

    print *, dlgama(c)
    if (abs(dlgama(c) - 1.95790392620801725e+02_dp) > 1e-12) error stop

    print *, dlgama(62.73971326260_dp)
    if (abs(dlgama(62.73971326260_dp) - 1.95790392620801725e+02_dp) > 1e-12) error stop

end program
```

The commented tests fail on linux.

**Comment by HarshitaKalani** [comment]
This test works perfectly on linux.
```fortran
program intrinsics_231
    use, intrinsic :: iso_fortran_env, only: dp => real64, sp => real32
    real(sp) :: y, z
    real(dp) :: a, b, c

    y = 23.82920_sp
    z = 62.739713_sp

    a = 9271.72826260_dp
    b = 23.82926260_dp
    c = 62.73971326260_dp

    print *, lgamma(y)
    if (abs(lgamma(y) - 5.10680733e+01) > 1e-5) error stop

    print *, lgamma(23.82920)
    if (abs(lgamma(23.82920) - 5.10680733e+01) > 1e-5) error stop

    print *, lgamma(z)
    if (abs(lgamma(z) - 1.95790405e+02) > 1e-5) error stop

    print *, lgamma(62.739713)
    if (abs(lgamma(62.739713) - 1.95790390e+02) > 1e-5) error stop

    print *, lgamma(a)
    if (abs(lgamma(a) - 7.54193119924598432e+04_dp) > 1e-12) error stop

    print *, lgamma(9271.72826260_dp)
    if (abs(lgamma(9271.72826260_dp) - 7.54193119924598432e+04_dp) > 1e-12) error stop

    print *, lgamma(b)
    if (abs(lgamma(b) - 5.10682662991165941e+01_dp) > 1e-12) error stop

    print *, lgamma(23.82926260_dp)
    if (abs(lgamma(23.82926260_dp) - 5.10682662991165941e+01_dp) > 1e-12) error stop

    print *, lgamma(c)
    if (abs(lgamma(c) - 1.95790392620801725e+02_dp) > 1e-12) error stop

    print *, lgamma(62.73971326260_dp)
    if (abs(lgamma(62.73971326260_dp) - 1.95790392620801725e+02_dp) > 1e-12) error stop

    print *, algama(y)
    if (abs(algama(y) - 5.10680733e+01) > 1e-5) error stop

    print *, algama(23.82920)
    if (abs(algama(23.82920) - 5.10680733e+01) > 1e-5) error stop

    print *, algama(z)
    if (abs(algama(z) - 1.95790405e+02) > 1e-5) error stop

    print *, algama(62.739713)
    if (abs(algama(62.739713) - 1.95790390e+02) > 1e-5) error stop

    print *, algama(a)
    if (abs(algama(a) - 7.54193119924598432e+04_dp) > 1e-12) error stop

    print *, algama(9271.72826260_dp)
    if (abs(algama(9271.72826260_dp) - 7.54193119924598432e+04_dp) > 1e-12) error stop

    print *, algama(b)
    if (abs(algama(b) - 5.10682662991165941e+01_dp) > 1e-12) error stop

    print *, algama(23.82926260_dp)
    if (abs(algama(23.82926260_dp) - 5.10682662991165941e+01_dp) > 1e-12) error stop

    print *, algama(c)
    if (abs(algama(c) - 1.95790392620801725e+02_dp) > 1e-12) error stop

    print *, algama(62.73971326260_dp)
    if (abs(algama(62.73971326260_dp) - 1.95790392620801725e+02_dp) > 1e-12) error stop

    print *, dlgama(a)
    if (abs(dlgama(a) - 7.54193119924598432e+04_dp) > 1e-12) error stop

    print *, dlgama(9271.72826260_dp)
    if (abs(dlgama(9271.72826260_dp) - 7.54193119924598432e+04_dp) > 1e-12) error stop

    print *, dlgama(b)
    if (abs(dlgama(b) - 5.10682662991165941e+01_dp) > 1e-12) error stop

    print *, dlgama(23.82926260_dp)
    if (abs(dlgama(23.82926260_dp) - 5.10682662991165941e+01_dp) > 1e-12) error stop

    print *, dlgama(c)
    if (abs(dlgama(c) - 1.95790392620801725e+02_dp) > 1e-12) error stop

    print *, dlgama(62.73971326260_dp)
    if (abs(dlgama(62.73971326260_dp) - 1.95790392620801725e+02_dp) > 1e-12) error stop

end program
```

**Comment by HarshitaKalani** [comment]
LMK if more information is required.

**Comment by Shaikh-Ubaid** [comment]
Thanks @HarshitaKalani !

In summary:

```fortran
! if (abs(algama(z) - 1.95790390e+02) > 1e-5) error stop     ! **The output for this on linux is: 1.95790405e+02**
```

and 

```fortran
! if (abs(lgamma(z) - 1.95790390e+02) > 1e-5) error stop   ! **The output for this on linux is: 1.95790405e+02**
```

do not work on LInux.

@parth121101 Do you have access to a Linux system? Would it be possible for you to look into why the lack of precision on Linux or how to increase/improve the precision on Linux?

**Comment by parth121101** [comment]
> Thanks @HarshitaKalani !
> 
> In summary:
> 
> ```fortran
> ! if (abs(algama(z) - 1.95790390e+02) > 1e-5) error stop     ! **The output for this on linux is: 1.95790405e+02**
> ```
> 
> and
> 
> ```fortran
> ! if (abs(lgamma(z) - 1.95790390e+02) > 1e-5) error stop   ! **The output for this on linux is: 1.95790405e+02**
> ```
> 
> do not work on LInux.
> 
> @parth121101 Do you have access to a Linux system? Would it be possible for you to look into why the lack of precision on Linux or how to increase/improve the precision on Linux?

@Shaikh-Ubaid as discussed I will remove this test right now and will be creating a new issue for the same.

File: integration_tests/intrinsics_231.f90 [hunk 1]
```diff
@@ -1,27 +1,23 @@
...
     y = 23.82920_sp
     z = 62.739713_sp
 
     a = 9271.72826260_dp
     b = 23.82926260_dp
     c = 62.73971326260_dp
 
-    print *, lgamma(x)
-    if (abs(lgamma(x) - 5710.34375) > 1e-5) error stop
-
```
Comment (Shaikh-Ubaid) [inline integration_tests/intrinsics_231.f90]: Just curious, I understand two other tests failed so we commented them out,  but why do we remove this test and the other related test below? Is it redundant?

**Comment by parth121101** [comment]
@Shaikh-Ubaid can we merge this as of now. I have created a dedicated issue for the failing tests.

**Review by certik — APPROVED** [review]
I think this looks good to me.

**Comment by certik** [comment]
@parth121101 when you merge PRs, please make sure the history is clean. If not, do a squash merge. In the merge above, you merged a commit https://github.com/lfortran/lfortran/pull/4066/commits/9c7d934498d1e99f46e937db4017a4fe41efb47f which adds git conflict resolution markers, which are then removed in subsequent commit. Such commits should not be merged.

**Comment by parth121101** [comment]
> @parth121101 when you merge PRs, please make sure the history is clean. If not, do a squash merge. In the merge above, you merged a commit [9c7d934](https://github.com/lfortran/lfortran/commit/9c7d934498d1e99f46e937db4017a4fe41efb47f) which adds git conflict resolution markers, which are then removed in subsequent commit. Such commits should not be merged.

Sure @certik I would take care about this.

**Pull request merged**